### ISSUE TITLE
Please add --pause flag to console runner

### ISSUE THREAD

**Issue by beautifulcoder** [nunit/nunit#1024]
The console runner documentation states there is `--pause` flag. But, this option has magically disappeared in the latest release candidate. I would like to be able to attach a debugger before tests run.

**Comment by CharliePoole** [comment]
It was removed in favor of --debug. Does that work for you? I'll fix the docs.

**Comment by beautifulcoder** [comment]
I guess I don't know how `--debug` works. `--pause` would wait for the debugger to attach, much like `dnx --debug`.

**Comment by rprouse** [comment]
The debug flag will bring up an attach to debugger window allowing you to
attach an existing or new instance of Visual Studio

**Comment by beautifulcoder** [comment]
I'll play with it again Monday. From what I remember this does not work. Open up a new issue?

**Comment by CharliePoole** [comment]
The advantage of --debug is that it will bring up the attach to debugger window from whatever process is running the test. It replaces and enhances --pause.

**Comment by beautifulcoder** [comment]
Just verified, `--debug` does not work on my machine with the latest release. Turning on tracing and verbose gives me this:

    NUnit Console Runner 3.0.5795
    Copyright (C) 2015 Charlie Poole

    Runtime Environment
       OS Version: Microsoft Windows NT 6.3.9600.0
      CLR Version: 4.0.30319.42000

    Test Files
        XXXXXXX

    Test Filters
        Where: XXXXXXX

    Agent process 10868 starting
    Running under version 4.0.30319.42000, Net 4.5
    Agent process 10868 exiting

    Test Run Summary
        Overall result: Passed
       Tests run: 169, Passed: 169, Errors: 0, Failures: 0, Inconclusive: 0
         Not run: 0, Invalid: 0, Ignored: 0, Explicit: 0, Skipped: 0
      Start time: 2015-11-16 14:20:40Z
        End time: 2015-11-16 14:20:42Z
        Duration: 1.822 seconds

Is this an issue?

**Comment by CharliePoole** [comment]
I have verified that it works using the released bits. Can you provide the exact command-line you are using?

**Comment by beautifulcoder** [comment]
Yes sir:
    >nunit3-console SomeProjectDLL.dll --debug --trace=Verbose -v --noresult --where="test==SomeNamespace && test!=SomeOtherClassName"

**Comment by CharliePoole** [comment]
I'm not seeing a problem. I ran exactly that command and the Visual Studio Just-In-Time debugger window popped up. What is your run environment? It's possible it may only fail for certain environments.

**Comment by beautifulcoder** [comment]
I'm on a work computer and the JIT debugger does not come up. My best guess is that it is a security issue on the machine. Is it possible to just bring `--pause` back? I much rather just attach to process, it is more intuitive for me.

**Comment by oznetmaster** [comment]
Why not have both?  They do different things.

**Comment by CharliePoole** [comment]
They **do** different things but they are intended **for** the same thing: to allow users to easily debug tests, including those run under an agent.

For the purpose, I think --debug is superior:

1. It automatically causes the break in the correct process - in the case of multiple agent processes, the user does not have to guess which one to attach.

2. It doesn't depend on the availabilty of any particular technology. As implemented, --pause depends on use of a Windows message box. (Naturally we could implement a more robust, cross-platform pause, but we chose to put the time into debug instead.)

Of course, this assumes that --debug actually works on the platform!

We could reinstate --pause for those who need it, but I'd rather put some time into figuring out exactly why --debug is not working for @beautifulcoder and documenting any pre-requisites for its use.

So... my first question is whether you have enabled just in time debugging for Visual Studio. I thought that an explicit call to launch the debugger would take precedence, but maybe it doesn't. If you haven't, please try enabling it and let us know the result.

**Comment by CharliePoole** [comment]
@beautifulcoder Also, please tell us about the environment: OS, VS level, NUnit version, etc.

**Comment by beautifulcoder** [comment]
I'm on Windows 8.1 Enterprise, VS 2015, NUnit 3.0.5797. Not sure about superior, like @oznetmaster said I believe in choice. I'm not an admin in my local dev environment at work. There might be a registry setting that disables the JIT debugger for security concerns. Personally, I much rather prefer to attach to process than get prompted. So why not:

    Console.WriteLine("Process ID: " + GetCurrentProcess().Id);
    while (!Debugger.IsAttached)
     {
         Thread.Sleep(100);
     }
     Console.WriteLine("Debugger attached");

Also, when `--pause` gets used I recommend turning on the `--inprocess` flag. When debugging, I'm not running through the entire suite of unit tests but a single one.

**Comment by CharliePoole** [comment]
Some observations on the side:

Developers need to have admin access to their own machines. This doesn't mean you always work using your admin id - you should have another id for normal development. As a consultant I have had to fight for that in many places - usually against the internal IT folks who want to control things. If there are security issues, build a wall around the dev group. Possibly out of your control, of course.

Back to your issue:

In order to work this issue adequately, we actually do need to know why --debug is not working for you. Is it possible you can find out for us? Then we can say something in the docs and possibly even give an error message if you can't do it.

We don't like to force changes in one setting due to another. The old pause worked quite well when running out of process, except you have to figure out what to attach to. If we add it back it will continue to work the same way.

If you use --inprocess and are running from VS you don't even need --pause. Just set a breakpoint and debug.

A --pause option that **requires** you to attach rather than just **allowing** it seems like a usability bug.

**Comment by beautifulcoder** [comment]
I have in good faith effort tried to figure out why `--debug` does not work. Usability aside, we do get local admin accounts but it can't be your login. I tried opening up a `cmd` with admin rights, run with `--trace`, and no juice. It just ignores the flag. The only thing I haven't done is check the event logs. I'll take another crack at it tomorrow and report findings, if any.

As a side note, I rather tell the tool to wait on me than be prompted. Who knows, the next time I want to attach might not be in Visual Studio so I like having this freedom.

**Comment by rprouse** [comment]
One potentially useful use of a pause command line option would be remote debugging. I've had cases where I want to debug a problem that only happens on a specific machine. If I could write, run and debug a quick test by only installing the remote debugger, it would have been very helpful.

The current --debug option might also work for that. It would be a good test to see if it allowed for attaching to the remote debugger.

**Comment by CharliePoole** [comment]
We'll try to get --pause back into the 3.0.1 releaese.

**Comment by CharliePoole** [comment]
@beautifulcoder You never replied as to whether you turned on JIT debugging, which is found under Tools | Options. I believe you must be running with admin privileges to turn it on or off, but not to actually use it. If you can, I suggest you at least try the option you are so dead-set against!

BTW, this feature allows selecting a debugger, which doesn't have to be Visual Studio. It also works with other IDEs that support it.

@rprouse I think this is easy enough to add back in and some users may see it as a loss. I put it into the 3.0.1 milestone, although not at a high priority. We should try to address remote debugging in 3.2.

**Comment by oznetmaster** [comment]
I see it as very useful in my environment, which is why my build still has it...

---Sent from Boxer | http://getboxer.com

On November 24, 2015 at 11:10:17 AM CST, CharliePoole <notifications@github.com> wrote:@beautifulcoder You never replied as to whether you turned on JIT debugging, which is found under Tools | Options. I believe you must be running with admin privileges to turn it on or off, but not to actually use it. If you can, I suggest you at least try the option you are so dead-set against!  BTW, this feature allows selecting a debugger, which doesn't have to be Visual Studio. It also works with other IDEs that support it.  @rprouse I think this is easy enough to add back in and some users may see it as a loss. I put it into the 3.0.1 milestone, although not at a high priority. We should try to address remote debugging in 3.2.  —Reply to this email directly or view it on GitHub.

**Comment by beautifulcoder** [comment]
@CharliePoole I do have the JIT debugger turned on. But I never get prompted for some odd reason. Checked event logs today, nothing. I'm not dead-set against it, It's just not an option in my environment. As a work around, I'm using `MDbg` to wrap the test tooling around a CLI debugger. I'm not proud but it works. Visual Studio crashes often with more than 80+ csproj in one solution. So anything that Visual Studio doesn't control is a win for me. I've experienced more stability when attaching vs firing the process from within. Anyway, I'm super stoked to see this on the roadmap!
